README.md
# update_kodi

NZBGet script to update Kodi's video library

## Options

* [host] => The IP address or resolvable hostname that Kodi is listening on.
* [port] => The port that Kodi is listening on for HTTP requests.
* [require_prior_scripts] => Whether or not prior scripts must succeed before updating Kodi.

## Limitations

* Requires Python 3.10+


CHANGELOG.md
# Changelog

## [v1.2.0](https://github.com/natemccurdy/nzbget-update_kodi/tree/v1.2.0)

[Full Changelog](https://github.com/natemccurdy/nzbget-update_kodi/compare/v1.1.2...v1.2.0)

**Implemented enhancements:**

- Add ability to run only if previous script succeeded [\#6](https://github.com/natemccurdy/nzbget-update_kodi/issues/6)

**Merged pull requests:**

- Various updates and fixes [\#7](https://github.com/natemccurdy/nzbget-update_kodi/pull/7) ([natemccurdy](https://github.com/natemccurdy))

## [v1.1.2](https://github.com/natemccurdy/nzbget-update_kodi/tree/v1.1.2) (2015-12-01)

[Full Changelog](https://github.com/natemccurdy/nzbget-update_kodi/compare/v1.1.1...v1.1.2)

**Fixed bugs:**

- Error messages aren't logging as ERROR [\#4](https://github.com/natemccurdy/nzbget-update_kodi/issues/4)
- update\_kodi.sh should be executable [\#3](https://github.com/natemccurdy/nzbget-update_kodi/issues/3)

**Merged pull requests:**

- Merge in changes for release 1.1.2 [\#5](https://github.com/natemccurdy/nzbget-update_kodi/pull/5) ([natemccurdy](https://github.com/natemccurdy))

## [v1.1.1](https://github.com/natemccurdy/nzbget-update_kodi/tree/v1.1.1) (2015-11-22)

[Full Changelog](https://github.com/natemccurdy/nzbget-update_kodi/compare/v1.1.0...v1.1.1)

**Closed issues:**

- Only executes if Kodi is running on the localhost [\#1](https://github.com/natemccurdy/nzbget-update_kodi/issues/1)

## [v1.1.0](https://github.com/natemccurdy/nzbget-update_kodi/tree/v1.1.0) (2015-11-22)

[Full Changelog](https://github.com/natemccurdy/nzbget-update_kodi/compare/v1.0.0...v1.1.0)

**Merged pull requests:**

- Release v1.1.0 [\#2](https://github.com/natemccurdy/nzbget-update_kodi/pull/2) ([natemccurdy](https://github.com/natemccurdy))

## [v1.0.0](https://github.com/natemccurdy/nzbget-update_kodi/tree/v1.0.0) (2015-07-18)

[Full Changelog](https://github.com/natemccurdy/nzbget-update_kodi/compare/f4dca63ca7d687cde0801277ff655ef1e816d3ce...v1.0.0)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


update_kodi.py
#!/usr/bin/env python3
##############################################################################
### NZBGET POST-PROCESSING SCRIPT                                          ###

# Update Kodi's video library
#
# This script sends a jsonrpc call to Kodi's API to update the video library.
#
# NOTE: Requires that Kodi's Web Interface is enabled: http://kodi.wiki/view/Web_interface
#
# NOTE: Requires Python 3+
#
#
# Web-site: http://github.com/natemccurdy/nzbget-update_kodi
#
# Version: 2.0.0

# fmt: off
##############################################################################
### OPTIONS                                                                ###

# The hostname or IP address of the host running Kodi.
#
# This can be a remote host or a local host. e.g. 192.168.1.50 or localhost
#host=127.0.0.1

# The port that Kodi is listening to for API calls.
#
# This should be the port number that you see on the 'Web Server' page in
# Kodi's settings.
#port=8081

# Require previous scripts to succeed before updating Kodi (yes, no).
#
# Setting this to 'yes' will check the exit status of all prior extension
# scripts. If any of them failed, Kodi will not be updated. This is useful when
# used in combination with VideoSort. For example, if VideoSort fails, there's
# no need to update Kodi. When this is set to 'yes', make sure to order this
# extension after VideoSort.
#require_prior_scripts=no
#
### NZBGET POST-PROCESSING SCRIPT                                          ###
##############################################################################
pass  # https://github.com/psf/black/issues/1245
# fmt: on

import os  # noqa: E402
import sys  # noqa: E402
from urllib.request import Request, urlopen  # noqa: E402
from urllib.error import HTTPError, URLError  # noqa: E402


SUCCESS = 93
FAILURE = 94
NONE = 95

HOST = os.getenv("NZBPO_HOST")
PORT = os.getenv("NZBPO_PORT")


def info(msg: str):
    print(f"[INFO] {msg}")


def warning(msg: str):
    print(f"[WARNING] {msg}")


def error(msg: str):
    print(f"[ERROR] {msg}")


def detail(msg: str):
    print(f"[DETAIL] {msg}")


def check_download_success():
    if os.getenv("NZBPP_TOTALSTATUS") != "SUCCESS":
        warning(f"The download of {os.getenv('NZBPP_NZBNAME')} has failed; skipping update.")
        sys.exit(NONE)


def check_prior_scripts():
    if os.getenv("NZBPP_SCRIPTSTATUS") != "SUCCESS":
        warning("Prior extension scripts did not succeed; skipping update.")
        sys.exit(NONE)


def main():
    check_download_success()
    if os.getenv("NZBPO_REQUIRE_PRIOR_SCRIPTS") == "yes":
        check_prior_scripts()

    kodi_url = f"http://{HOST}:{PORT}/jsonrpc"
    rpc_command = '{ "jsonrpc": "2.0", "method": "VideoLibrary.Scan", "id": "mybash"}'
    timeout = 5

    request = Request(url=kodi_url, data=rpc_command.encode("utf-8"))
    request.add_header("Content-Type", "application/json; charset=utf-8")

    detail(f"Sending Kodi video library update to: {request.full_url}")

    try:
        with urlopen(request, timeout=timeout) as response:
            result = response.read().decode()
    except HTTPError as e:
        error(f"Can't connect to {request.full_url} - {e.code} - {e.reason}")
        sys.exit(FAILURE)
    except URLError as e:
        error(f"Can't connect to {request.full_url} - {e.reason}")
        sys.exit(FAILURE)
    else:
        info(f"Kodi video library update successful - {result}")
        sys.exit(SUCCESS)


if __name__ == "__main__":
    main()


update_kodi.sh
#!/bin/bash
##############################################################################
### NZBGET POST-PROCESSING SCRIPT                                          ###

# Update Kodi's video library.
#
# This script sends a jsonrpc call to Kodi's API to update the video library.
#
# It also requires that the web interface be enabled in Kodi. Instructions
# for that can be found here:
#     http://kodi.wiki/view/Web_interface
#
# Info about update_kodi:
# Web-site: http://github.com/natemccurdy/nzbget-update_kodi.
# PP-Script Version: 1.1.2.
#
# NOTE: This script only runs on *nix based hosts with BASH.
#       It also requires that curl is installed and is in the $PATH.

##############################################################################
### OPTIONS                                                                ###

# The hostname or IP address of the host running Kodi.
#
# This can be a remote host or a local host. e.g. 192.168.1.50 or localhost
#host=127.0.0.1

# The port that Kodi is listening to for API calls.
#
# This should be the port number that you see on the 'Web Server' page in
# Kodi's settings.
#port=8081

# Whether to force an update or not (yes, no).
#
# If Kodi is on the same host as NZBGet, we can check to see that Kodi is
# running before trying to hit its API. This way, we get a clean "Skip" in
# NZBGET instead of an "ERROR". If you want to disable this check and
# always try to hit the API, set this to 'yes'.
#force_update=no

# Require previous scripts to succeed before updating Kodi (yes, no).
#
# Setting this to 'yes' will check the exit status of all prior extension
# scripts. If any of them failed, Kodi will not be updated. This is useful when
# used in combination with VideoSort. For example, if VideoSort fails, there's
# no need to update Kodi. When this is set to 'yes', make sure to order this
# extension after VideoSort.
#require_prior_scripts=no
#
### NZBGET POST-PROCESSING SCRIPT                                          ###
##############################################################################

SUCCESS=93
ERROR=94
SKIP=95

# Check that the required options have been set before continuing
[[ -n $NZBPO_HOST ]] || { echo "[ERROR] Host not set"; exit $ERROR; }
[[ -n $NZBPO_PORT ]] || { echo "[ERROR] Port not set"; exit $ERROR; }

kodi_is_local () {
  hostname="$(hostname)"
  case "$NZBPO_HOST" in
    localhost|127.0.0.1|"$hostname") return 0;;
    *)                               return 1;;
  esac
}

kodi_is_running_locally () {
  if pgrep -i kodi 1>/dev/null 2>&1; then
    return 0
  else
    return 1
  fi
}

if [[ $NZBPP_TOTALSTATUS != SUCCESS ]]; then
  echo "[WARNING] The download of ${NZBPP_NZBNAME} has failed; skipping update."
  exit $SKIP
fi

if [[ $NZBPO_REQUIRE_PRIOR_SCRIPTS == yes ]]; then
  if [[ $NZBPP_SCRIPTSTATUS != SUCCESS ]]; then
    echo "[WARNING] Prior extension scripts did not succeed; skipping update."
    exit $SKIP
  fi
fi

if [[ $NZBPO_FORCE_UPDATE == no ]]; then
  if kodi_is_local && ! kodi_is_running_locally; then
    echo "[WARNING] Kodi is not running so we can't update it; skipping update."
    exit $SKIP
  fi
fi

if ! command -v curl 1>/dev/null 2>&1; then
  # shellcheck disable=SC2016
  echo '[ERROR] Can not find curl. update_kodi requires curl to be installed and in $PATH.'
  exit $ERROR
fi

curl --connect-timeout 5 \
  --data-binary \
  '{ "jsonrpc": "2.0", "method": "VideoLibrary.Scan", "id": "mybash"}' \
  -H 'content-type: application/json;' \
  "http://${NZBPO_HOST}:${NZBPO_PORT}/jsonrpc" 1>/dev/null 2>&1

curl_return_value="$?"

case $curl_return_value in
  0)
    exit $SUCCESS ;;
  6)
    echo "[ERROR] Couldn't resolve host: ${NZBPO_HOST}"
    exit $ERROR ;;
  7)
    echo "[ERROR] Could not connect to the Kodi API endpoint at ${NZBPO_HOST}:${NZBPO_PORT}."
    echo "[ERROR] Is Kodi running and is 'Allow remote control via HTTP' enabled?"
    exit $ERROR ;;
  *)
    echo "[ERROR] Unknown error occured. Curl returned: ${curl_return_value}"
    exit $ERROR ;;
esac
